classes/task/reminder.php

```diff
@@ -0,0 +1,77 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Reminder task
+ *
+ * @package    mod_kanban
+ * @copyright  2023 ISB Bayern
+ * @author     Stefan Hanauska
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+namespace mod_kanban\task;
+
+use mod_kanban\helper;
+
+/**
+ * Reminder task
+ *
+ * @package    mod_kanban
+ * @copyright  2023 ISB Bayern
+ * @author     Stefan Hanauska
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class reminder extends \core\task\scheduled_task {
+
+    /**
+     * Return the task's name as shown in admin screens.
+     *
+     * @return string
+     */
+    public function get_name() {
+        return get_string('remindertask', 'mod_kanban');
+    }
+
+    /**
+     * Execute the task.
+     */
+    public function execute() {
+        global $DB;
+        $kanbancards = $DB->get_records_sql(
+            'SELECT ' . $DB->sql_concat('c.id', '"-"', 'a.userid') . ' as uniqid,
+                    c.id as id, c.title as title, k.name as boardname, c.duedate as duedate, a.userid as userid, k.id as instance
+             FROM {kanban_card} c
+             JOIN {kanban_assignee} a ON a.kanban_card = c.id
+             JOIN {kanban_board} b ON b.id = c.kanban_board
+             JOIN {kanban} k ON b.kanban_instance = k.id
+             WHERE c.duedate != 0
```

Comment:
All the where conditions are only targeting table `kanban_card`, so the where clause should be directly after `FROM {kanban_card} c` to first reduce the amount of records which afterwards will be joined with all the other tables. In case the underlying database layer or DBMS will not do this optimization by itself the query could be sped up like that probably.

Reply:
Moving the WHERE clause does not comply with the SQL standards - but I moved these query parts into the on clause of the first join (also for some other queries), so this should be done first.